github: Added matrix strategies for linux workflows.

marcfehling opened the pull request:
Part of #17583.

I found this branch sitting in my repository. Worked on it a while ago but forgot about it. Might as well merge it now. 

This patch used to update most of the github workers, but since we already updated some in the past there is not much left of it after a rebase.

masterleinad commented:
What's the motivation for updating at this point?

marcfehling commented:
> What's the motivation for updating at this point?

No particular motivation, just a general update.

Personally, I would have one less stale branch in my repository :smile:

bangerth commented:
CI isn't my area, but I would like to point out that there is value in not running all CI steps on the same base OS version. We catch more bugs with different OS versions, compiler versions, etc.

marcfehling commented:
> CI isn't my area, but I would like to point out that there is value in not running all CI steps on the same base OS version. We catch more bugs with different OS versions, compiler versions, etc.

We can run those workflows on multiple OSs which use compilers that come with the OS, i.e., `linux release serial`, `linux debug parallel`, `linux debug parallel tpetra`, `OSX serial`, `OSX parallel 64bit`. The Windows workflows actually do this already; other workflows use compilers that we set up on the fly. I can adjust this PR accordingly.

masterleinad commented:
Sure we can try adding more configurations. Let's see how much more expensive that gets.

tjhei commented:
> 24 22 20

github just announced that 20.04 runners will go away in April

marcfehling commented:
I have added matrix strategies for most linux and osx workflows. For this I added consistent names for each job that reflects the OS they are using. I adjusted the other name accordingly.

While there, I got rid of redundancies in the naming -- this is fine as github fetches information through the yml files (so for example our badges about the status of the CI on the landing page will be fine).

marcfehling commented on .github/workflows/linux.yml:
```diff
@@ -42,8 +42,17 @@ jobs:
+        - os: ubuntu-22.04
+          image: dealii/dependencies:jammy
+        - os: ubuntu-24.04
+          image: dealii/dependencies:jammy
```
We don't have a `noble` image yet. But the `jammy` image will do for now.

masterleinad reviewed (approved):
Looks good to me.

gassmoeller commented on .github/workflows/linux.yml:
```diff
@@ -24,15 +24,15 @@ env:
+  ##################
 
-  linux-release-serial:
+  release-serial:
```
I dont understand why you use a matrix strategy for runners for this job and for `debug-parallel` and `debug-parallel-tpetra` (every job that contains `container: image:`). These jobs are executed inside the container we specify in `image: ...`. So which runner we use outside the container doesnt matter. Dont you duplicate these runs with the matrix strategy you propose here?

I would understand if you want to use a matrix strategy for different images, but as you mention we dont have different images (or at least dont use them at the moment).

marcfehling replied:
> So which runner we use outside the container doesnt matter. Dont you duplicate these runs with the matrix strategy you propose here?

I admit that my understanding of Docker containers is limited. You are right -- it makes no sense to run a container image (see also [here](https://github.com/orgs/community/discussions/25534#discussioncomment-3248244)) on two different host OSs. Until we have a `noble` docker container, my suggestion does not make sense.

gassmoeller commented on .github/workflows/docker.yml:
```diff
@@ -1,4 +1,4 @@
-name: github-docker
```
I think the `github` part was to separate these runs visually from the `Jenkins: ...` workflows. I dont feel particularly strongly about it, but wouldnt `Github: ...` be more similar to the other workflows?

marcfehling replied:
In my opinion the little icon is enough to identify github workflows -- at least on the website. On notification emails about failing CI, it might be still be nice to immediately see that a github runner failed.

I am open to rename them to `Github: <...>`. The main purpose of my suggestion was to reduce the length of the complete string that identifies a runner.

gassmoeller reviewed (commented):
I am all for improvements of the CI workflows, but I have two questions about this PR. Can you let me know if I misunderstood something there?

marcfehling commented:
We can continue to use the docker images or fall back to dealii backports for github actions. We do not have the docker image for our dependencies on ubuntu noble yet. I would like to help, but my experience with docker is very limited. Do we have instructions on how to set it up using [dealii/docker-files](https://github.com/dealii/docker-files)?

gassmoeller commented:
Like I said above I am happy to review and merge improvements to the CI, but I think this PR is currently a bit problematic, because it is hard to answer the question: What is the improvement in this PR? It mixes a number of changes:
- renaming the workflows
- adding matrix builds on different runners for several workflows (some are useful, like the ones for Macos and Windows, but the ones running the linux docker containers are not useful until we have new docker images)
- updating the runner versions of some workflows

Can you separate the changes into independent PRs?

My general feeling is:
- Renaming the workflows is a matter of personal taste, I am fine either way
- Introducing matrix builds wherever they actually test different OSs makes sense, so keep those changes in this PR
- Using different operating systems for other builds (like the `linux-debug-intel-oneapi` workflow) 

> We can continue to use the docker images or fall back to dealii backports for github actions. We do not have the docker image for our dependencies on ubuntu noble yet. I would like to help, but my experience with docker is very limited.

I think a mixture of dealii backports and docker images is actually the best. Like Wolfgang wrote we want to test a host of different configurations. Everything that needs to measure accuracy should be run inside the docker images, but we also want to test compiling on different platforms.

> Do we have instructions on how to set it up using [dealii/docker-files](https://github.com/dealii/docker-files)?

Not really, I was struggling with the same thing when I tried looking into creating the arm images a while ago. It looks like the images on dockerhub are built by a bot called [dealiibot](https://hub.docker.com/u/dealiibot), but I dont know who is running that bot. @luca-heltai is that you?

masterleinad commented:
> Do we have instructions on how to set it up using [dealii/docker-files](https://github.com/dealii/docker-files?rgh-link-date=2025-02-25T12%3A35%3A11Z)?

It should be enough to run
```
docker build -t dealii/dependencies:noble --build-arg VERSION=9.6.0-1~ubuntu24.04.1~ppa1 --build-arg IMG=noble --build-arg REPO=ppa:ginggs/deal.ii-9.6.0-backports --build-arg CLANG_VERSION=16 --build-arg CLANG_REPO=https://github.com/dealii/dealii/releases/download/v9.6.0/ .
```
in the `dependencies` directory and push the result to Docker Hub.

masterleinad commented:
> @masterleinad Let me try tomorrow.

I gave it a shot, see https://github.com/dealii/docker-files/pull/46.

marcfehling commented:
> they had announced that they would also provide arm builders

The arm workers are currently on a "public preview". See [github doc](https://docs.github.com/en/actions/using-github-hosted-runners/using-github-hosted-runners/about-github-hosted-runners#standard-github-hosted-runners-for-public-repositories) and [github blog](https://github.blog/changelog/2025-01-16-linux-arm64-hosted-runners-now-available-for-free-in-public-repositories-public-preview/). We can tackle to automate building the combined deal.II amd64/arm64 images completely in github (in a separate PR).

> For the dependencies this needs to be done manually. I'll do it as soon as I get back home.

This does not have a high priority!

masterleinad commented:
Looks like `masterleinad/dependencies-dealii:noble` is working well for now.

luca-heltai commented on .github/workflows/linux.yml:
```diff
@@ -42,8 +42,17 @@ jobs:
+        - os: ubuntu-22.04
+          image: dealii/dependencies:jammy
+        - os: ubuntu-24.04
+          image: masterleinad/dependencies-dealii:noble
```
```suggestion
          image: dealii/dependencies:noble
```
This is now up-to-date

luca-heltai commented:
@masterleinad there is a manual worflow in docker-files that rebuilds and push the dependencies to the repository. 

I'm running it now here:

https://github.com/dealii/docker-files/actions/runs/13562352647/job/37907855021

The change you made for noble (installing clang-format-16) broke jammy, as there is no such package for jammy... We don't need to fix it right away, as the jammy image is still good, but it would be nice if we could support the last two stable versions of ubuntu throughout.

masterleinad commented:
> The change you made for noble (installing clang-format-16) broke jammy, as there is no such package for jammy... We don't need to fix it right away, as the jammy image is still good, but it would be nice if we could support the last two stable versions of ubuntu throughout.

We could just use https://github.com/dealii/dealii/blob/master/contrib/utilities/download_clang_format instead.

masterleinad commented:
> You mean first wgetting it from the repo, and then downloading clang?

Yes, exactly.

luca-heltai commented:
> > You mean first wgetting it from the repo, and then downloading clang?
> 
> Yes, exactly.

Ok. See https://github.com/dealii/docker-files/pull/47
